Fix dockerng.network_* name matching

abulford opened the pull request:
Fixes #41982

### What does this PR do?

This PR is a copy of #41988 but against the 2017.7 branch.  The rest of the description below is copied from there.

Updates the states so that they loop through the list of networks returned from `dockerng.networks` to check that one of the networks returned fully matches the name of the network that needs to be ensured to be present or absent.

Another option would be to add an argument to the `dockerng.networks` execution module to enable strict matching.  This would probably be preferable to repeating the logic in the states, but I didn't want to meddle too much as the execution module is currently just wrapping the docker-py library.

### What issues does this PR fix or reference?
#41982

### Previous Behavior
`dockerng.network_present` would not create a network if another network with a partially matching name already existed, `docker.network_absent` would try to remove a network which didn't exist if there was another network in existence with a partially matching name.

```
user@host ~ $ sudo docker network ls | grep cde
e982840fa7a1        abcdef              bridge              local
user@host ~ $ sudo salt '*' state.apply docker.network.test_present test=True
host.domain:
----------
          ID: test_network_present
    Function: dockerng.network_present
        Name: cde
      Result: True
     Comment: Network 'cde' already exists.
     Started: 15:14:34.886654
    Duration: 5.399 ms
     Changes:   

Summary for host.domain
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   5.399 ms
user@host ~ $ sudo salt '*' state.apply docker.network.test_present
host.domain:
----------
          ID: test_network_present
    Function: dockerng.network_present
        Name: cde
      Result: True
     Comment: Network 'cde' already exists.
     Started: 15:14:41.984847
    Duration: 9.433 ms
     Changes:   

Summary for host.domain
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   9.433 ms
user@host ~ $ sudo salt '*' state.apply docker.network.test_absent test=True
host.domain:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: None
     Comment: The network 'cde' will be removed
     Started: 15:16:26.964736
    Duration: 4.956 ms
     Changes:   

Summary for host.domain
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   4.956 ms
user@host ~ $ sudo salt '*' state.apply docker.network.test_absent
host.domain:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: False
     Comment: Failed to remove network 'cde': Error 404: {"message":"network cde not found"}
     Started: 15:16:14.349035
    Duration: 9.475 ms
     Changes:   

Summary for host.domain
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   9.475 ms
ERROR: Minions returned with non-zero exit code
```

### New Behavior
The `dockerng.network_present` and `dockerng.network_absent` states correctly recognise whether a network is present or absent, even if another network with a partially matching name exists:
```
user@host ~ $ sudo docker network ls | grep cde
e982840fa7a1        abcdef              bridge              local
user@host ~ $ sudo salt '*' state.apply .docker.network.test_present test=True
host.domain.com:
----------
          ID: test_network_present
    Function: dockerng.network_present
        Name: cde
      Result: None
     Comment: The network 'cde' will be created
     Started: 16:21:02.715381
    Duration: 4.757 ms
     Changes:   

Summary for host.domain.com
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   4.757 ms
user@host ~ $ sudo salt '*' state.apply .docker.network.test_present
host.domain.com:
----------
          ID: test_network_present
    Function: dockerng.network_present
        Name: cde
      Result: True
     Comment: 
     Started: 16:21:41.720032
    Duration: 501.139 ms
     Changes:   
              ----------
              created:
                  ----------
                  Id:
                      246079aa13211fe681949b2f6f5066c2b425e631645296e078a0acff4fafbf9a
                  Warning:

Summary for host.domain.com
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 501.139 ms
user@host ~ $ sudo docker network ls | grep cde
e982840fa7a1        abcdef              bridge              local
246079aa1321        cde                 bridge              local
user@host ~ $ sudo salt '*' state.apply .docker.network.test_absent test=True
host.domain.com:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: None
     Comment: The network 'cde' will be removed
     Started: 16:21:52.552362
    Duration: 5.432 ms
     Changes:   

Summary for host.domain.com
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   5.432 ms
user@host ~ $ sudo salt '*' state.apply .docker.network.test_absent
host.domain.com:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: True
     Comment: 
     Started: 16:22:09.832561
    Duration: 383.447 ms
     Changes:   
              ----------
              removed:
                  None

Summary for host.domain.com
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 383.447 ms
user@host ~ $ sudo docker network ls | grep cde
e982840fa7a1        abcdef              bridge              local
user@host ~ $ sudo salt '*' state.apply .docker.network.test_absent test=True
host.domain.com:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: True
     Comment: Network 'cde' already absent
     Started: 16:22:18.913120
    Duration: 5.008 ms
     Changes:   

Summary for host.domain.com
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   5.008 ms
user@host ~ $ sudo salt '*' state.apply .docker.network.test_absent
host.domain.com:
----------
          ID: test_network_absent
    Function: dockerng.network_absent
        Name: cde
      Result: True
     Comment: Network 'cde' already absent
     Started: 16:22:24.240611
    Duration: 5.547 ms
     Changes:   

Summary for host.domain.com
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   5.547 ms
```

### Tests written?
No

abulford committed:
Fix dockerng.network_* name matching

Updating the states so that they loop through the list of networks
returned from `dockerng.networks` to check that one of the networks
returned fully matches the name of the network that needs to be ensured
to be present or absent.

Another option would be to add an argument to the `dockerng.networks`
execution module to enable strict matching.  This would probably be
preferable to repeating the logic in the states, but I didn't want to
meddle too much as the execution module is currently just wrapping the
docker-py library.

Fixes #41982

salt-jenkins commented:
@abulford, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage and @rallytime to be potential reviewers.

rallytime commented:
@abulford The same test is failing on this change as well: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/12448/ 

Don't worry about the other one though - only the docker test is related. :)

rallytime commented:
Go Go Jenkins!

rallytime commented:
Go Go Jenkins!

rallytime commented:
@abulford The related test that is failing here is `unit.states.test_docker_network.DockerNetworkTestCase.test_absent`. Can you take a look at that? Once that test failure is resolved, we can get this merged in.

Please let us know if you need any help.

abulford commented:
@rallytime sorry for the delayed response, I just got back from holiday.

I see in #41988 you've said that the error is unrelated to this change - before I delve back into this to see what's going on, do you know if this is also the case for this PR?

rallytime commented:
@abulford No worries! I hope you had a nice holiday. :)

The test I mentioned above is related to this PR. I checked out your PR and I can reproduce that failure, even after rebasing your changes on top of the current HEAD of 2017.7. The test in question here is `unit.states.test_docker_network.DockerNetworkTestCase.test_absent`. Please let me know if you have any questions.

rallytime commented:
@abulford Also, looking at #41988, it looks like you already have a fix there for that particular test. Perhaps that was supposed to be applied here as well?

abulford commented:
Duh, I've just totally missed applying the same fix to this branch.  Sorry @rallytime!  Will get that done first thing tomorrow (UK time).

abulford committed:
Add trace logging of docker.networks result

Added during debugging of failing test to determine what
docker.networks is expected to return, keeping as it seems useful.

abulford committed:
Fix broken unit test test_network_absent

This started failing following commit 515c612, which relied on the
'Name' key being present in the return value of docker.networks -
as the mock didn't have this set the test started failing.

abulford committed:
Add regression tests for #41982

These test the scenarios where another network with a similar name
already exists, verifying that absent doesn't attempt to remove a
network which isn't specified, and present still attempts to create the
specified network despite a similarly named network already being present.

abulford commented:
@rallytime I've applied the changes from #41988 to this branch, with tweaks to account for function name changes.

Commit:
Lint: Remove extra line at end of file.

rallytime commented:
Thanks @abulford! I fixed a small lint issue. Once tests are done running we can get this in.

abulford commented:
@rallytime Sorry about that, but thanks for fixing :).

rallytime commented:
@abulford Oh, no problem at all. It was easy to just grab it. Thank you again for your contributions and work here. It is very much appreciated. :)

abulford commented:
No problem @rallytime, thanks for merging :).
